jupyterhub/_xsrf_utils.py

```diff
@@ -230,6 +234,17 @@ def check_xsrf_cookie(handler):
         )
 
 
+# allow disabling using ip in anonymous id
+_anonymous_use_ip = _bool_env("JUPYTERHUB_XSRF_ANONYMOUS_USE_IP", default=True)
```

Comment:
Shouldn't we use `JUPYTERHUB_XSRF_ANONYMOUS_USE_IP` with False by default to avoid any security problem and keep the old behaviour?

I think this problem is more likely to occur in k8s deployment, so we can set this property to True on the chart side.

Reply:
I don't think so. True means keeping the current behavior by default, where the ip is used, which is correct most of the time (including in most z2jh deployments). Only where proxy headers don't properly propagate does this problem come up.

Reply:
Doesn't this default change the behaviour on internal networks, since all private IPs are treated as one?

Reply:
Yes, it does. I'm not particularly concerned about that case, since it only affects the non-portability of the xsrf cookies for login, i.e. for cookie tossing scenarios. I don't think that would really be an issue for private network deployments.

Reply:
I still don't feel comfortable with treating all private IPs as one, but I'm not going to block merging this if others are happy.

As one possible alternative could we replace `JUPYTERHUB_XSRF_ANONYMOUS_USE_IP` with something like `JUPYTERHUB_XSRF_ANONYMOUS_<SOMETHING>_CIDRS` which defaults to empty (so no change to current mainline behaviour), but setting it to a CIDR would treat any IP in that CIDR as identical for the XSRF check? Then the equivalent of `JUPYTERHUB_XSRF_ANONYMOUS_USE_IP=False` would be `JUPYTERHUB_XSRF_ANONYMOUS_<SOMETHING>_CIDRS="0.0.0.0/0"`

Reply:
I implemented your suggestion as `JUPYTERHUB_XSRF_ANONYMOUS_IP_CIDRS`. When a CIDR matches, the matching CIDR is used in the hash instead of the ip (so e.g. access from multiple flattened networks is still non-transferrable across networks), and it is now opt-in.

That means https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/3422 is no longer fixed by default, but it is _fixable_ and we may be able to make it fixed by default in z2jh, I _think_.